fix(compiler-cli): fix type narrowing of `@if` with aliases

alxhub opened the pull request:
When an `@if` expression has an alias, only the type of the alias is currently narrowed. So for example, suppose `value` is `string|undefined`:

```
@if (value; as alias) {
  {{ value.length }} <!-- error, value may be undefined -->
  {{ alias.length }} <!-- no error, alias is narrowed -->
}
```

This is especially noticeable when the expression contains guards which are preconditions for the aliased expression:

```
@if (a && b; as alias) {...}
```

In this case, `a` would not be narrowed within the body, even though the `@if` condition forces it to be truthy. This is a bug.

The reason is that aliased expressions were previously type-checked as:

```
var alias = a && b;
if (alias) {
  // nothing other than alias is narrowed
  ...
}
```

One option considered was to emit `const alias` instead of `var alias`. TypeScript _does_ trace `const` expressions and narrow their individual components when the overall expression is guarded:

```
const alias = a && b;
if (alias) {
  // a, b are also narrowed
}
```

However, this narrowing has different semantics than if `a && b` appeared directly in the guard expression. For example, object properties aren't narrowed with this approach, so component properties (which are referenced as e.g. `this.a`) would not be narrowed.

Instead, a different technique is used: an alias variable is declared inside the `if` guard, capturing the inferred type.

```
if (a && b) {
  var alias = a && b;
  // a, b, and alias all narrowed correctly.
}
```

This form ensures all conditions within the guard expression get narrowed while also narrowing the alias variable type, and works for property narrowing as well. It also allows renaming to rename the alias variable, since all references to that variable are generated with correct source mapping.

Fixes #52855

kbrilla commented:
quick quiestioin in this aproach a && b if functions will be called twice right, actually the whole expression will be called twice. 
Why not do something like this? 
<img width="449" alt="image" src="https://github.com/angular/angular/assets/22288449/d3fbf24e-022f-447c-82db-1c90b86e9982">
<img width="353" alt="image" src="https://github.com/angular/angular/assets/22288449/4247de6f-1806-4a90-9dfc-800e8af77ed3">

This way we would get alias and original expression narrowed and do comparision once. 

### More examples with difrent usages

```html
@if (a.type === 'success')
```

<img width="640" alt="image" src="https://github.com/angular/angular/assets/22288449/e87a7325-c28a-4203-9d84-a9b9c312e1c0">

```html
@if (a?.optional?.data; as alias) {
....
}
```
<img width="367" alt="image" src="https://github.com/angular/angular/assets/22288449/d6a6daa4-4cfc-4f4c-995c-e32fbece12be">

crisbeto commented:
@kbrilla this code only runs during template type checking, it doesn't affect how many times it'll be called at runtime.

kbrilla commented:
so possibly this also fixes my old issue https://github.com/angular/angular/issues/47077, with recomendation of `@if` usage instead of `*ngIf`

alxhub committed:
fix(compiler-cli): fix type narrowing of `@if` with aliases

When an `@if` expression has an alias, only the type of the alias is
currently narrowed. So for example, suppose `value` is `string|undefined`:

```
@if (value; as alias) {
  {{ value.length }} <!-- error, value may be undefined -->
  {{ alias.length }} <!-- no error, alias is narrowed -->
}
```

This is especially noticeable when the expression contains guards which are
preconditions for the aliased expression:

```
@if (a && b; as alias) {...}
```

In this case, `a` would not be narrowed within the body, even though the
`@if` condition forces it to be truthy. This is a bug.

The reason is that aliased expressions were previously type-checked as:

```
var alias = a && b;
if (alias) {
  // nothing other than alias is narrowed
  ...
}
```

One option considered was to emit `const alias` instead of `var alias`.
TypeScript _does_ trace `const` expressions and narrow their individual
components when the overall expression is guarded:

```
const alias = a && b;
if (alias) {
  // a, b are also narrowed
}
```

However, this narrowing has different semantics than if `a && b` appeared
directly in the guard expression. For example, object properties aren't
narrowed with this approach, so component properties (which are referenced
as e.g. `this.a`) would not be narrowed.

Instead, we amend the guard expression to include both the expression _and_ the
alias variable, enforcing that both are narrowed.

```
var alias = a && b;
if ((a && b) && alias) {
  // a, b, and alias all narrowed correctly.
}
```

This form ensures all conditions within the guard expression get narrowed
while also narrowing the alias variable type.

Fixes #52855

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1661,15 +1665,18 @@ class TcbIfOp extends TcbOp {
-        markIgnoreDiagnostics(expression);
-      } else {
-        expression = expressionScope.resolve(branch.expressionAlias);
+      // We need to recreate the expression and mark it to be ignored for diagnostics,
+      // because it was already checked as a part of the block's condition and we don't
+      // want it to produce a duplicate diagnostic.
+      expression = tcbExpression(branch.expression, this.tcb, expressionScope);
+      if (branch.expressionAlias !== null) {
+        expression = ts.factory.createBinaryExpression(
+          ts.factory.createParenthesizedExpression(expression),
+          ts.SyntaxKind.AmpersandAmpersandToken,
+          expressionScope.resolve(branch.expressionAlias),
+        );
```
not blocking - This looks like the exact same code block as above. Can it be refactored to be a shared function?

alxhub commented:
Caretaker: mergeability check seems frozen, but should be good to go.

dylhunn commented:
This PR was merged into the repository by commit 9884875c9698028c5788ad150239e4c52109b87d.
